refactored/src/entry_table.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::{HashMap, HashSet};
+
+use cosmian_crypto_core::{
+    kdf,
+    reexport::rand_core::{RngCore, SeedableRng},
+    symmetric_crypto::{
+        aes_256_gcm_pure::{
+            decrypt_in_place_detached, encrypt_in_place_detached, MAC_LENGTH, NONCE_LENGTH,
+        },
+        key::Key,
+        SymKey,
+    },
+    CsRng,
+};
+
+use crate::{
+    callbacks::{FetchEntry, UpsertEntry},
+    edx::Edx,
+    error::Error,
+};
+
+/// Encrypted value contained inside the Entry Table. It is composed of the
+/// AESGCM-256 encrypted value, the nonce used and the corresponding MAC tag.
+pub struct EncryptedValue<const VALUE_LENGTH: usize> {
+    ciphertext: [u8; VALUE_LENGTH],
+    tag: [u8; MAC_LENGTH],
+    nonce: [u8; NONCE_LENGTH],
+}
+
+impl<const VALUE_LENGTH: usize> EncryptedValue<VALUE_LENGTH> {
+    fn encrypt(
+        key: &[u8],
+        value: [u8; VALUE_LENGTH],
+    ) -> Result<Self, cosmian_crypto_core::CryptoCoreError> {
+        let mut res = Self {
+            ciphertext: value,
+            nonce: [0; NONCE_LENGTH],
+            tag: [0; MAC_LENGTH],
+        };
+        let mut rng = CsRng::from_entropy();
```

Comment:
Maybe accept a `CsRng` in trait parameters to avoid creating new one during each encrypt?

Reply:
I moved the CsRng into the `Index` struct to avoid generating a new one in each call to `add` or `delete` (which the old version was doing).